File: app/src/main/java/com/tablehockey/game/model/TeamInfo.kt
package com.tablehockey.game.model

import android.graphics.Color
import java.io.Serializable

/**
 * A selectable club. All names are fictional so the game can ship on the
 * Amazon Appstore without any league or club trademarks.
 */
data class TeamInfo(
    val city: String,
    val name: String,
    val abbr: String,
    val primary: Int,
    val secondary: Int,
    val text: Int,
    val league: String = LEAGUE_PRO
) : Serializable {
    val fullName: String get() = if (city.isEmpty()) name else "$city $name"

    companion object {
        const val LEAGUE_PRO = "Pro League"
        const val LEAGUE_CALGARY = "Timbits U7"

        /**
         * Calgary-area Timbits (U7) clubs, one per minor hockey association.
         * Colours are approximations; edit here to match real jerseys.
         */
        private val CALGARY: List<TeamInfo> = listOf(
            TeamInfo("Glenlake", "Hawks", "GLK", Color.parseColor("#0B6E3A"), Color.parseColor("#F5D130"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Bow Valley", "Flames", "BVF", Color.parseColor("#C8102E"), Color.parseColor("#F1BE48"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Crowfoot", "Coyotes", "CRO", Color.parseColor("#1C2E5A"), Color.parseColor("#C4432B"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Blackfoot", "Chiefs", "BFC", Color.parseColor("#7A1F1F"), Color.parseColor("#F3E5AB"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Springbank", "Rockies", "SPB", Color.parseColor("#4B2E83"), Color.parseColor("#C0C0C0"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Trails West", "Wolves", "TWW", Color.parseColor("#1F2937"), Color.parseColor("#14B8A6"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("McKnight", "Mustangs", "MCK", Color.parseColor("#1E3A8A"), Color.parseColor("#F59E0B"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Northwest", "Warriors", "NWW", Color.parseColor("#14532D"), Color.parseColor("#FDE047"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Simons Valley", "Storm", "SVS", Color.parseColor("#0F3D6E"), Color.parseColor("#9CA3AF"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Bow River", "Bruins", "BRB", Color.parseColor("#111111"), Color.parseColor("#FCB514"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Midnapore", "Mavericks", "MID", Color.parseColor("#7F1D1D"), Color.parseColor("#111827"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Lake Bonavista", "Breakers", "LBB", Color.parseColor("#1E40AF"), Color.parseColor("#7DD3FC"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Southwest", "Cougars", "SWC", Color.parseColor("#B91C1C"), Color.parseColor("#111827"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Crowchild", "Blackhawks", "CCB", Color.parseColor("#B91C1C"), Color.parseColor("#0B0B0B"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Calgary", "Knights", "KNI", Color.parseColor("#2F3E5C"), Color.parseColor("#E5E7EB"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Airdrie", "Lightning", "AIR", Color.parseColor("#1D4ED8"), Color.parseColor("#FACC15"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Okotoks", "Oilers", "OKO", Color.parseColor("#0F2A5C"), Color.parseColor("#F97316"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Chestermere", "Lakers", "CHE", Color.parseColor("#0E7490"), Color.parseColor("#FDE68A"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Foothills", "Flyers", "FTH", Color.parseColor("#EA580C"), Color.parseColor("#111827"), Color.WHITE, LEAGUE_CALGARY),
            TeamInfo("Strathmore", "Storm", "STR", Color.parseColor("#374151"), Color.parseColor("#60A5FA"), Color.WHITE, LEAGUE_CALGARY)
        )

        val ALL: List<TeamInfo> = CALGARY + listOf(
            TeamInfo("Denver", "Peaks", "DEN", Color.parseColor("#1D4ED8"), Color.parseColor("#F8FAFC"), Color.WHITE),
            TeamInfo("Chicago", "Blizzard", "CHI", Color.parseColor("#B91C1C"), Color.parseColor("#FBBF24"), Color.WHITE),
            TeamInfo("Montreal", "Royals", "MTL", Color.parseColor("#7C3AED"), Color.parseColor("#FDE68A"), Color.WHITE),
            TeamInfo("Vancouver", "Orcas", "VAN", Color.parseColor("#0F766E"), Color.parseColor("#E2E8F0"), Color.WHITE),
            TeamInfo("Boston", "Harbor", "BOS", Color.parseColor("#111827"), Color.parseColor("#F59E0B"), Color.WHITE),
            TeamInfo("Minnesota", "Timber", "MIN", Color.parseColor("#166534"), Color.parseColor("#FCA5A5"), Color.WHITE),
            TeamInfo("Dallas", "Longhorns", "DAL", Color.parseColor("#0E7490"), Color.parseColor("#F1F5F9"), Color.WHITE),
            TeamInfo("Toronto", "Pilots", "TOR", Color.parseColor("#1E3A8A"), Color.parseColor("#93C5FD"), Color.WHITE)
        )

        fun byIndex(i: Int): TeamInfo = ALL[i.coerceIn(0, ALL.size - 1)]

        /** Index of the default home club (Glenlake Hawks). */
        val DEFAULT_HOME: Int = ALL.indexOfFirst { it.city == "Glenlake" }.coerceAtLeast(0)
        val DEFAULT_AWAY: Int = ALL.indexOfFirst { it.city == "Bow Valley" }.coerceAtLeast(1)

        /** Spinner label: club name plus its league. */
        fun label(t: TeamInfo): String = t.fullName + "  ·  " + t.league
    }
}

File: ios/PowerPlayHockeyCore/Sources/PowerPlayHockeyCore/Model/TeamInfo.swift
/// A selectable club. All names are fictional so the game can ship without
/// any league or club trademarks.
struct TeamInfo {
    static let LEAGUE_PRO = "Pro League"
    static let LEAGUE_CALGARY = "Timbits U7"

    let city: String
    let name: String
    let abbr: String
    let primary: UInt32
    let secondary: UInt32
    let text: UInt32
    let league: String

    init(_ city: String, _ name: String, _ abbr: String, _ primary: UInt32, _ secondary: UInt32, _ text: UInt32, _ league: String = TeamInfo.LEAGUE_PRO) {
        self.city = city
        self.name = name
        self.abbr = abbr
        self.primary = primary
        self.secondary = secondary
        self.text = text
        self.league = league
    }

    var fullName: String { city.isEmpty ? name : "\(city) \(name)" }

    /// Calgary-area Timbits (U7) clubs, one per minor hockey association.
    /// Colours are approximations; edit here to match real jerseys.
    private static let CALGARY: [TeamInfo] = [
        TeamInfo("Glenlake", "Hawks", "GLK", HexColor.argb("#0B6E3A"), HexColor.argb("#F5D130"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Bow Valley", "Flames", "BVF", HexColor.argb("#C8102E"), HexColor.argb("#F1BE48"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Crowfoot", "Coyotes", "CRO", HexColor.argb("#1C2E5A"), HexColor.argb("#C4432B"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Blackfoot", "Chiefs", "BFC", HexColor.argb("#7A1F1F"), HexColor.argb("#F3E5AB"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Springbank", "Rockies", "SPB", HexColor.argb("#4B2E83"), HexColor.argb("#C0C0C0"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Trails West", "Wolves", "TWW", HexColor.argb("#1F2937"), HexColor.argb("#14B8A6"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("McKnight", "Mustangs", "MCK", HexColor.argb("#1E3A8A"), HexColor.argb("#F59E0B"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Northwest", "Warriors", "NWW", HexColor.argb("#14532D"), HexColor.argb("#FDE047"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Simons Valley", "Storm", "SVS", HexColor.argb("#0F3D6E"), HexColor.argb("#9CA3AF"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Bow River", "Bruins", "BRB", HexColor.argb("#111111"), HexColor.argb("#FCB514"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Midnapore", "Mavericks", "MID", HexColor.argb("#7F1D1D"), HexColor.argb("#111827"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Lake Bonavista", "Breakers", "LBB", HexColor.argb("#1E40AF"), HexColor.argb("#7DD3FC"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Southwest", "Cougars", "SWC", HexColor.argb("#B91C1C"), HexColor.argb("#111827"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Crowchild", "Blackhawks", "CCB", HexColor.argb("#B91C1C"), HexColor.argb("#0B0B0B"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Calgary", "Knights", "KNI", HexColor.argb("#2F3E5C"), HexColor.argb("#E5E7EB"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Airdrie", "Lightning", "AIR", HexColor.argb("#1D4ED8"), HexColor.argb("#FACC15"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Okotoks", "Oilers", "OKO", HexColor.argb("#0F2A5C"), HexColor.argb("#F97316"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Chestermere", "Lakers", "CHE", HexColor.argb("#0E7490"), HexColor.argb("#FDE68A"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Foothills", "Flyers", "FTH", HexColor.argb("#EA580C"), HexColor.argb("#111827"), HexColor.white, LEAGUE_CALGARY),
        TeamInfo("Strathmore", "Storm", "STR", HexColor.argb("#374151"), HexColor.argb("#60A5FA"), HexColor.white, LEAGUE_CALGARY)
    ]

    static let ALL: [TeamInfo] = CALGARY + [
        TeamInfo("Denver", "Peaks", "DEN", HexColor.argb("#1D4ED8"), HexColor.argb("#F8FAFC"), HexColor.white),
        TeamInfo("Chicago", "Blizzard", "CHI", HexColor.argb("#B91C1C"), HexColor.argb("#FBBF24"), HexColor.white),
        TeamInfo("Montreal", "Royals", "MTL", HexColor.argb("#7C3AED"), HexColor.argb("#FDE68A"), HexColor.white),
        TeamInfo("Vancouver", "Orcas", "VAN", HexColor.argb("#0F766E"), HexColor.argb("#E2E8F0"), HexColor.white),
        TeamInfo("Boston", "Harbor", "BOS", HexColor.argb("#111827"), HexColor.argb("#F59E0B"), HexColor.white),
        TeamInfo("Minnesota", "Timber", "MIN", HexColor.argb("#166534"), HexColor.argb("#FCA5A5"), HexColor.white),
        TeamInfo("Dallas", "Longhorns", "DAL", HexColor.argb("#0E7490"), HexColor.argb("#F1F5F9"), HexColor.white),
        TeamInfo("Toronto", "Pilots", "TOR", HexColor.argb("#1E3A8A"), HexColor.argb("#93C5FD"), HexColor.white)
    ]

    static func byIndex(_ i: Int) -> TeamInfo {
        ALL[min(max(i, 0), ALL.count - 1)]
    }

    /// Index of the default home club (Glenlake Hawks).
    static let DEFAULT_HOME: Int = max(ALL.firstIndex { $0.city == "Glenlake" } ?? 0, 0)
    static let DEFAULT_AWAY: Int = max(ALL.firstIndex { $0.city == "Bow Valley" } ?? 1, 1)

    /// Picker label: club name plus its league.
    static func label(_ t: TeamInfo) -> String { t.fullName + "  ·  " + t.league }
}
